packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,99 @@
+const ERROR_MESSAGE_BYTES = (value) => `${ERROR_MESSAGE_PREFIX} bytes "${value}".`;
+const ERROR_MESSAGE_STRING = (value) => `${ERROR_MESSAGE_PREFIX} string "${value}".`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], rawValues: any[] = []): string {
+  if(types.length !== rawValues.length) throw new Error('Supplied number of types and values do not match.');
+  const values = [];
+  _.zipWith(types, rawValues, function(type: string, value: any) {
+    values.push(parseTypeValuePair(type, value));
+  });
+  const methodId = abi.methodID(name, types).toString('hex');
+  const params = abi.rawEncode(types, values).toString('hex');
   return '0x' + methodId + params;
 }
+
+export function parseTypeValuePair(type: string, rawValue: any): string | never {
+  if(type === 'address') {
+    return parseAddress(rawValue);
+  }
+  else if(type === 'string') {
+    return rawValue; // Validated by ethereumjs-abi.
+  }
+  else if(type.startsWith('bytes') && !type.includes('[]')) {
+    return parseBytes(rawValue);
+  }
+  else if(type.startsWith('uint') && !type.includes('[]')) {
+    return parseNumber(rawValue, true);
+  }
+  else if(type.startsWith('int') && !type.includes('[]')) {
+    return parseNumber(rawValue, false);
+  }
+  else {
+    // TODO: parse remianing types: fixed, arrays, etc.
+    return rawValue;
+  }
+}
+
+function parseBytes(value: any): string | never {
+  if(typeof(value) !== 'string') throw new Error(ERROR_MESSAGE_BYTES(value));
+  if(value.toString().length === 0) return value;
+  if(!/^(0x)?[0-9a-f]$/i.test(value)) throw new Error(ERROR_MESSAGE_BYTES(value));
+  return value;
+}
+
+function parseAddress(value: any): string | never {
+  if(typeof(value) !== 'string') throw new Error(ERROR_MESSAGE_ADDRESS(value));
+  if(!/^(0x)?[0-9a-f]{40}$/i.test(value)) throw new Error(ERROR_MESSAGE_ADDRESS(value));
+  // TODO: It'd be nice to perform a checksum validation here.
+  return value;
+}
+
+function parseNumber(value: any, mustBePositive: boolean): string | never {
+  if(typeof(value) === 'number') return parseNumberFromLiteral(value, mustBePositive);
+  else if(typeof(value) === 'string') return parseNumberFromString(value, mustBePositive);
+  else if(BN.isBigNumber(value)) return parseNumberFromBigNumber(new BN(value), mustBePositive);
+  throw new Error(ERROR_MESSAGE_NUMBER(value));
+}
+
+function parseNumberFromBigNumber(value: BN, mustBePositive: boolean): string | never {
+  if(mustBePositive && value.isNegative()) throw new Error(ERROR_MESSAGE_POSITIVE_NUMBER(value));
+  if(!value.isInteger()) throw new Error(ERROR_MESSAGE_INTEGER_NUMBER(value));
+  return value.toString();
```

Comment:
Can we add a check to ensure that value is not NaN, which was the original issue?

Reply:
Yes, but I'm not understanding what you said about "which was the original issue". The original issue was caused by our scientific notation regex identifying an address with 'e' characters as a scientific notation number.

Reply:
Sort-of. The original issue was caused by that regex, but also because of the fact that we happily sent `NaN`s as valid integer values without checking them. I'd like to fix both issues.

Reply:
In that case, yes, we are now doing isNaN checks with regular Js as well as with BN.isNaN.